pkg/disttask/framework/taskexecutor/manager.go

```diff
@@ -294,39 +317,55 @@ func (m *Manager) cancelTaskExecutors(tasks []*proto.TaskBase) {
 	m.mu.RLock()
 	defer m.mu.RUnlock()
 	for _, task := range tasks {
-		m.logger.Info("cancelTasks", zap.Int64("task-id", task.ID))
+		m.logger.Info("cancel task executor", zap.Int64("task-id", task.ID))
 		if executor, ok := m.mu.taskExecutors[task.ID]; ok {
 			executor.Cancel()
 		}
 	}
 }
 
 // startTaskExecutor handles a runnable task.
-func (m *Manager) startTaskExecutor(taskBase *proto.TaskBase) {
+func (m *Manager) startTaskExecutor(taskBase *proto.TaskBase) (executorStarted bool) {
 	// TODO: remove it when we can create task executor with task base.
 	task, err := m.taskTable.GetTaskByID(m.ctx, taskBase.ID)
 	if err != nil {
 		m.logger.Error("get task failed", zap.Int64("task-id", taskBase.ID), zap.Error(err))
-		return
+		return false
+	}
+	if !m.slotManager.alloc(&task.TaskBase) {
+		m.logger.Info("alloc slots failed, maybe other task executor alloc more slots at runtime",
+			zap.Int64("task-id", taskBase.ID), zap.Int("concurrency", taskBase.Concurrency),
+			zap.Int("remaining-slots", m.slotManager.availableSlots()))
+		return false
 	}
-	// runCtx only used in executor.Run, cancel in m.fetchAndFastCancelTasks.
+	defer func() {
+		// free the slot if executor not started.
```

Comment:
otherwise line 373 will free it? Can we unify the free in this `defer`?

Reply:
yes, but we cannot unify them. this line is for handling TaskExecutor creation errors, else slots should be freed after TaskExecutor exits, like the `Close` if `Init` failed pattern.